base/linalg/matmul.jl

```diff
@@ -656,6 +656,7 @@ function matmul2x2{T,S}(tA, tB, A::AbstractMatrix{T}, B::AbstractMatrix{S})
 end
 
 function matmul2x2!{T,S,R}(C::AbstractMatrix{R}, tA, tB, A::AbstractMatrix{T}, B::AbstractMatrix{S})
+    size(C) == (2, 2) || throw(DimensionMismatch("Destination matrix is not 2×2"))
```

Comment:
Updated. I just copied the checking code from `Base.BLAS.gemm!` so hopefully that should everything entirely consistent.